### PR TITLE
Add flexbox, third, quarter, three quarters columns, grow and stretch…

### PR DESCRIPTION
**Pull request by ayumihamsaki**
Fixes this github issue: https://github.com/octobercms/october/issues/4437

### CODE REVIEW

**Comment by ayumihamsaki** [comment]
Notes:

Many people have suggested to use pure CSS, but have done little research before writing there suggestions! There is no pure css solution and this pull request has been tested on many browsers and even IE11 support.

Because of this, the offset's are calculated using javascript. A pure CSS solution would be to use Flexbox `space-evenly`, this is a no go! There is an Edge bug for this: https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/15947692/

Plus there will never be any support for Flexbox `space-evenly` in IE11.

Another issue would be that trying to use a pure CSS method would also not work for Safari browsers below version 10.1

This pull request has had many hours of researching behind it to get a multi-browser support solution. Please understand this before commenting.

**Comment by ayumihamsaki** [comment]
Two-third test results:

image

**Comment by ayumihamsaki** [comment]
## Added line-break feature

### Example 1 (No Line-breaks)

```markdown
fields:
    one:
        label: 'Two-thirds'
        span: two-thirds
        type: text
    two:
        label: 'Quarter'
        span: 'quarter'
        type: text
    three:
        label: 'Third'
        span: third
        type: text
    four:
        label: 'Third'
        span: third
        type: text
    five:
        label: 'Third'
        span: third
        type: text
```

Gives us the following result:

image

### Example 2 (Let's make the Quarter field have a Line-break)

```markdown
fields:
    one:
        label: 'Two-thirds'
        span: two-thirds
        type: text
    two:
        label: 'Quarter'
        span: 'quarter line-break'
        type: text
    three:
        label: 'Third'
        span: third
        type: text
    four:
        label: 'Third'
        span: third
        type: text
    five:
        label: 'Third'
        span: third
        type: text
```

Gives us the following result:

image

**Comment by ayumihamsaki** [comment]
Previous commit fixes the offset issue with a line-break so now gives the result:

image

- The middle `third` is now placed correctly in the center.

**Comment by ayumihamsaki** [comment]
=== Working on line-break feature right now ===

**Comment by ayumihamsaki** [comment]
Finished.

**Comment by ayumihamsaki** [comment]
Re: Admins if you want to add a Milestone to this pull request. I'm keen to get it merged it whenever everyone's free. Just want to try and avoid any develop branch conflicts due to so many files been updated. Done a huge amount of testing time and there shouldn't be any errors or very few.

File: modules/backend/formwidgets/codeeditor/partials/_codeeditor.htm [hunk 1]
```diff
@@ -5,7 +5,7 @@
 <?php else: ?>
     <div
         id="<?= $this->getId() ?>"
-        class="field-codeeditor size-<?= $size ?> <?= $stretch?'layout-relative':'' ?>"
+        class="field-codeeditor size-<?= $size ?> <?= $stretch?'layout-relative':'' ?> <?= $grow?'layout-relative':'' ?>"
```
[the remark this answers is not in the document]
Comment (ayumihamsaki) [reply modules/backend/formwidgets/codeeditor/partials/_codeeditor.htm]: fixed

File: modules/backend/formwidgets/markdowneditor/partials/_markdowneditor.htm [hunk 2]
```diff
@@ -3,7 +3,7 @@
 <?php else: ?>
     <div
         id="<?= $this->getId() ?>"
-        class="field-markdowneditor size-<?= $size ?> <?= $stretch?'layout-relative stretch':'' ?>"
+        class="field-markdowneditor size-<?= $size ?> <?= $stretch?'layout-relative stretch':'' ?> <?= $grow?'layout-relative grow':'' ?>"
```
[the remark this answers is not in the document]
Comment (ayumihamsaki) [reply modules/backend/formwidgets/markdowneditor/partials/_markdowneditor.htm]: fixed

File: modules/backend/formwidgets/richeditor/assets/less/richeditor.less [hunk 3]
```diff
@@ -54,14 +54,6 @@
 @btn-active-text: #000;
 @input-label-color: @color-label;
 
-//
-// Make the focus ring and textarea fill the whole rich editor container
-//
-
-.fr-element {
-    height: 100%;
-}
```
[the remark this answers is not in the document]
Comment (ayumihamsaki) [reply modules/backend/formwidgets/richeditor/assets/less/richeditor.less]: Yah it's being doubled, link: https://github.com/octobercms/october/pull/4485/files

(will remove it)

[the remark this answers is not in the document]
Comment (ayumihamsaki) [reply modules/backend/formwidgets/richeditor/assets/less/richeditor.less]: @LukeTowers Going to take a screenshot - as something is not right your end. This is gone already and this comment is on the left hand side for some strange reason?

Code already moved to here: https://github.com/octobercms/october/pull/4485/file

image

File: modules/backend/formwidgets/richeditor/assets/less/richeditor.less [hunk 4]
```diff
@@ -137,10 +129,10 @@
         .fr-wrapper {
             width: 100% !important;
             left: 0;
-            top: 0;
+            top: 0 !important;
             bottom: 0;
-            position: absolute;
-            height: auto !important;
+            position: relative;
+            height: 100% !important;
```
[the remark this answers is not in the document]
Comment (ayumihamsaki) [reply modules/backend/formwidgets/richeditor/assets/less/richeditor.less]: These changes are intended. I can't find any other commit that has changed this.

**Comment by ayumihamsaki** [comment]
Please dont merge this until I have finished creating a pr for the test plugin and remove an annoying issue with auto stretch in the tabs.

**Comment by ayumihamsaki** [comment]
Please dont merge this until I have finished creating a pr for the test plugin.

### Re-test everything in test plugin

- [x] - New grid fields - passed testing

- [x] - New line fields - passed testing

Grow tests:
- [ ] - Grow - not passing
  - [x] Textarea
  - [x] Rich editor
  - [ ] Markdown editor
  - [ ] Code editor

Stretch tests:
- [ ] - Stretch - not passing (In hidden tabs)

**Comment by ayumihamsaki** [comment]
Please dont merge this until I have finished creating a pr for the test plugin.

### Re-test everything in test plugin

- [x] - New grid fields - passed testing

- [x] - New line fields - passed testing

Grow tests:

- [ ] - Grow - not passing - see below
  - [x] Textarea
  - [x] Rich editor
  - [x] Markdown editor
  - [x] Code editor

There is a bug with the grow on the Markdown editor and Code editor widgets, it sometimes doesn't work and some times does, without changing any code. This strange bug seems to be from using an old version of Ace Editor.

Stretch tests:

- [ ] - Stretch - not passing (In hidden tabs)

Progress notes: https://github.com/ayumihamsaki/grow-stretch-temp/blob/master/README.md

**Comment by ayumihamsaki** [comment]
@LukeTowers @bennothommo 

Hit a small problem now, I need the following pr's tested and merged:

- Merge Order 1: Add Tab WAI-ARIA Support https://github.com/octobercms/october/pull/4401 

- Merge Order 2: Add keyboard navigation and focus ring for radio buttons , slider switches and list fields (this pr udates the tab styling to allow the focus ring to work on the tabs) https://github.com/octobercms/october/pull/4501

- Content tabs do not work at all in the backend - this makes them now work https://github.com/octobercms/october/pull/4506

Basically these three PR's fix the tabs in October.

I need to calculate the height of the tabs against the height of the form area to work out how much I can stretch a field. So I can't finish off this PR with the stretch function until those PR's are merged and then I got a stable height styling for the tabs to use.

Hope that makes sense?

**Comment by ayumihamsaki** [comment]
@LukeTowers @bennothommo 

Hit a small problem now, I need the following pr's tested and merged:

- Merge Order 1: Add Tab WAI-ARIA Support https://github.com/octobercms/october/pull/4401 

- Merge Order 2: Add keyboard navigation and focus ring for radio buttons , slider switches and list fields (this pr udates the tab styling to allow the focus ring to work on the tabs) https://github.com/octobercms/october/pull/4501

- Content tabs do not work at all in the backend - this makes them now work https://github.com/octobercms/october/pull/4506

Basically these three PR's fix the tabs in October.

I need to calculate the height of the tabs against the height of the form area to work out how much I can stretch a field. So I can't finish off this PR with the stretch function until those PR's are merged and then I got a stable height styling for the tabs to use.

Hope that makes sense?

e.g.

image

**Comment by ayumihamsaki** [comment]
Updated test plugin (to make testing easier), found here: https://github.com/octoberrain/test-plugin/pull/77

**Comment by ayumihamsaki** [comment]
@LukeTowers @bennothommo 

> This PR is now completely finished - there is nothing more I need to do now!

- This PR: https://github.com/octobercms/october/pull/4509 (finished)

- Doc's PR: https://github.com/octobercms/docs/pull/387 (finished)

- Test plugin PR: https://github.com/octoberrain/test-plugin/pull/77 (finished)

Please install this PR and use the Test plugin PR I created and you should see the following results:

### Grid

image

### New line

image

### Grow

image

### Stretch (textarea)

image

### Stretch (rich editor)

image

### Stretch (markdown editor)

image

### Stretch (Code editor)

image

### Stretch (Hint)

image

### Stretch (Partial)

image

## Test results in test plugin

- [x] - New grid fields - passed testing (Stable)

- [x] - New line fields - passed testing (Stable)

Grow tests:

- [ ] - Grow - not passing - see below
  - [x] Textarea (Stable)
  - [x] Rich editor (Stable)
  - [x] Markdown editor (Unstable due to Ace version)
  - [x] Code editor (Unstable due to Ace version)

There is a bug with the grow on the Markdown editor and Code editor widgets, it sometimes doesn't work and some times does, without changing any code. This strange bug seems to be from using an old version of Ace Editor. This PR might fix all these issues found here: [Update Ace Editor from 1.2.6 to 1.4.5 and emmet.js](https://github.com/octobercms/october/pull/4585)

Stretch tests:

- [x] - Stretch - passed testing (Stable)
  - [x] Textarea (Stable)
  - [x] Rich editor (Stable)
  - [x] Markdown editor (Stable)
  - [x] Code editor (Stable)
  - [x] Hint (Stable)
  - [x] Partial (Stable)

**Comment by ayumihamsaki** [comment]
@LukeTowers @bennothommo 

> This PR is now completely finished - there is nothing more I need to do now!

- This PR: https://github.com/octobercms/october/pull/4509 (finished)

- Doc's PR: https://github.com/octobercms/docs/pull/387 (finished)

- Test plugin PR: https://github.com/octoberrain/test-plugin/pull/77 (finished)

Please install this PR and use the Test plugin PR I created and you should see the following results:

### Grid

image

### New line

image

### Grow

image

### Stretch (textarea)

image

### Stretch (rich editor)

image

### Stretch (markdown editor)

image

### Stretch (Code editor)

image

### Stretch (Hint)

image

### Stretch (Partial)

image

## Test results in test plugin

- [x] - New grid fields - passed testing (Stable)

- [x] - New line fields - passed testing (Stable)

Grow tests:

- [ ] - Grow - not passing - see below
  - [x] Textarea (Stable)
  - [x] Rich editor (Stable)
  - [x] Markdown editor (Unstable due to Ace version)
  - [x] Code editor (Unstable due to Ace version)

There is a bug with the grow on the Markdown editor and Code editor widgets, it sometimes doesn't work and some times does, without changing any code. This strange bug seems to be from using an old version of Ace Editor. This PR might fix all these issues found here: [Update Ace Editor from 1.2.6 to 1.4.5 and emmet.js](https://github.com/octobercms/october/pull/4585)

Stretch tests:

- [x] - Stretch - passed testing (Stable)
  - [x] Textarea (Stable)
  - [x] Rich editor (Stable)
  - [x] Markdown editor (Stable)
  - [x] Code editor (Stable)
  - [x] Hint (Stable)
  - [x] Partial (Stable)

**Comment by SebastiaanKloos** [comment]
@bennothommo is there something I can do to help this getting merged?

**Comment by LukeTowers** [comment]
@SebastiaanKloos you could test this trying to break it as much as possible and report back on the results

**Comment by SebastiaanKloos** [comment]
I've tested the PR. Overall I think it really benefits the FormController. This is the only issue I found out so far. The buttons are padded too much

CleanShot 2019-10-30 at 11 29 18

**Comment by SebastiaanKloos** [comment]
I will create a new PR for each function

**Comment by daftspunk** [comment]
The overall idea is for this is good, however, this PR is a bit of a mess and nearly impossible to review. If we're going to be adding more granular controls over the columns it makes sense to align with the Bootstrap column system that is already implemented. Overall, the solution also appears inelegant and depends on JavaScript when it shouldn't have to do this.
